aptos-move/framework/aptos-stdlib/sources/cryptography/multi_ed25519.move

```diff
@@ -0,0 +1,231 @@
+    //
+
+    /// An *unvalidated*, k out of n MultiEd25519 public key. The `bytes` field contains (1) several chunks of
+    /// `ed25519::PUBLIC_KEY_NUM_BYTES` bytes, each encoding a Ed25519 PK, and (2) a single byte encoding the threshold k.
+    /// *Unvalidated* means there is no guarantee that the underlying PKs are valid elliptic curve points of non-small
+    /// order.
+    struct UnvalidatedPublicKey has copy, drop, store {
+        bytes: vector<u8>
+    }
+
+    /// A *validated* k out of n MultiEd25519 public key. *Validated* means that all the underlying PKs will be
+    /// elliptic curve points that are NOT of small-order. It does not necessarily mean they will be prime-order points.
+    /// This struct encodes the public key exactly as `UnvalidatedPublicKey`.
+    ///
+    /// For now, this struct is not used in any verification functions, but it might be in the future.
+    struct ValidatedPublicKey has copy, drop, store {
+        bytes: vector<u8>
+    }
+
+    /// A purported MultiEd25519 multi-signature that can be verified via `verify_signature_strict` or
+    /// `verify_signature_strict_t`. The `bytes` field contains (1) several chunks of `ed25519::SIGNATURE_NUM_BYTES`
+    /// bytes, each encoding a Ed25519 signature, and (2) a `BITMAP_NUM_OF_BYTES`-byte bitmap encoding the signer
+    /// identities.
+    struct Signature has copy, drop, store {
+        bytes: vector<u8>
+    }
+
+    //
+    // Functions
+    //
+
+    /// Parses the input 32 bytes as an *unvalidated* MultiEd25519 public key.
+    public fun new_unvalidated_public_key_from_bytes(bytes: vector<u8>): UnvalidatedPublicKey {
+        assert!(vector::length(&bytes) % INDIVIDUAL_PUBLIC_KEY_NUM_BYTES == THRESHOLD_SIZE_BYTES, error::invalid_argument(E_WRONG_PUBKEY_SIZE));
+        UnvalidatedPublicKey { bytes }
+    }
+
+    /// Parses the input bytes as a *validated* MultiEd25519 public key.
+    public fun new_validated_public_key_from_bytes(bytes: vector<u8>): Option<ValidatedPublicKey> {
+        assert!(vector::length(&bytes) % INDIVIDUAL_PUBLIC_KEY_NUM_BYTES == THRESHOLD_SIZE_BYTES, error::invalid_argument(E_WRONG_PUBKEY_SIZE));
+        if(public_key_validate_internal(bytes)) {
+            option::some(ValidatedPublicKey {
+                bytes
+            })
+        } else {
+            option::none<ValidatedPublicKey>()
+        }
+    }
+
+    /// Parses the input bytes as a purported MultiEd25519 multi-signature.
+    public fun new_signature_from_bytes(bytes: vector<u8>): Signature {
+        assert!(vector::length(&bytes) % INDIVIDUAL_SIGNATURE_NUM_BYTES == BITMAP_NUM_OF_BYTES, error::invalid_argument(E_WRONG_SIGNATURE_SIZE));
+        Signature { bytes }
+    }
+
+    /// Converts a ValidatedPublicKey to an UnvalidatedPublicKey, which can be used in the strict verification APIs.
+    public fun public_key_to_unvalidated(pk: &ValidatedPublicKey): UnvalidatedPublicKey {
+        UnvalidatedPublicKey {
+            bytes: pk.bytes
+        }
```

Comment:
Huh? Why?

Reply:
validated -> unvalidated, why?

Reply:
Oh, we have a similar API in Ed25519.

For now, the signature verification APIs work with `UnvalidatedPublicKey`'s (because we validate them in the actual signature verification).

I wanted to make sure developers have a way of easily converting a `ValidatePublicKey` into an `UnvalidatedPublicKey` for the purpose of using this API.

I tend to agree, we don't need it though: one could get the bytes of the validate PK and reconstruct an unvalidated one (which is what this function does).